### ISSUE TITLE
Assigning async options

### ISSUE THREAD

**Issue by eleftrik** [monterail/vue-multiselect#215]
With Vue 2 version when I make an Ajax request, I set options to response.data.

```
multiselect(
  :options="countries",
  :selected="selectedCountries",
  :multiple="multiple",
  :searchable="searchable",
  @search-change="asyncFind",
  placeholder="Type to search",
  label="name"
  key="code"
)
  span(slot="noResult").
    Oops! No elements found. Consider changing the search query.
```

```
methods: {
  asyncFind (query) {
            axios.post(this.ajaxUrl, this.ajaxParams)
                .then(response => {
                    this.options = response.data; // <---- the problem is here
                })
  }
}
```

Of course, I got this error:

> [Vue warn]: Avoid mutating a prop directly since the value will be overwritten whenever the parent component re-renders. Instead, use a data or computed property based on the prop's value. Prop being mutated: "options" 

In your example, you do `this.countries = findService(query)`, but I think it's the same approach.

How can I do? 

Thanks

**Comment by eleftrik** [comment]
Maybe I managed it.

When my custom component receives an Ajax response, it emits a custom event, let's say `clientAjaxChanged`:

```
methods: {
  asyncFind (query) {
		this.isLoading = true;
		axios.post(this.ajaxUrl, this.ajaxParams)
			.then(response => {
				this.$emit('ajaxchanged', response.data);
				this.isLoading = false;
			})

  }
}

```

HMTL:

`<my-component :options="myOptions" @ajaxchanged="clientAjaxChanged"></my-component>
`

JS:

```
            methods: {
                clientAjaxChanged: function(values) {
                    if (values && values !== undefined) {
                        this.myOptions = values;
                    }
                },

```

It seems to work; I would only like to know if this approach is correct or if there is a better way.

Thanks

**Comment by eleftrik** [comment]
@shentao, any suggestion?

**Comment by shentao** [comment]
Sorry. Didn't had the time to properly analyze your issue just until now. :)

Well, first of – please change `:selected` to `:value` for Vue 2.x. If you pass the options from your container component, let’s say the structure looks like this:
```html
<container>
  <my-wrapper-component :options="myOptions">
    <multiselect>
    </multiselect>
  </my-wrapper-component>
</container>
```
then you have to update `myOptions` inside the `<container>`, not in the <my-wrapper-component>`.  Looking at your example you’re passing the options to the wrapper from the container, but you’re also running the async request inside the wrapper component, that’s why you get the `props being mutated` error. So either do not pass the options from the container to the wrapper or handle the search inside the container rather than inside the wrapper.

The way you did it now should work however, I’m not sure if this is a good idea to shuffle the data like this: 
```
multiselect@seach-change -> 
  wrapper’s asyncFind() call -> 
    emit @ajaxchanged and send the response -> 
      receive the response and update the options inside the container -> 
        it gets passed down to the multiselect
```

I guess you could simplify it by doing the async request and updating the option inside the container and only use the wrapper to forward the options down to the multiselect and the search-change event up to the container.

Would that make sense in your case?

**Comment by eleftrik** [comment]
Hi @shentao, thanks for your reply and for your time. Yes, In my code I used `:value`, here it was only a copy & paste from a GitHub example of yours.
Sorry, I didn't explain myself properly, I'm reading my post again and it's confusing. So, I try to explain better.

Yes, I have a wrapper component (`my-component`), which encapsulates your vue-multiselect.
I would like to put all the Ajax logic inside it, because the actions around `@search-change` are always the same. When the user types something in a `my-component` instance, this is a simplified list of the actions it should do:

- make an Ajax call to an URL (passed as a `prop`), passing what the user typed in
- take the result (array of objects, always in the form `{id: n, label: 'string'}`) and store it in the options prop

Here 2 examples (both working) which, I hope, clarify what I'm trying to explain :-)

Example 1: https://jsfiddle.net/eleftrik/aLn42w5q/
In this one, the wrapping component alters whe `options` property, but this operation raises a Vue warning

Example 2: https://jsfiddle.net/eleftrik/c94ks3Lj/
In this one, the wrapping component emits a custom event; in the container, this event is handled and the array passed to :options (cities) can be changed without warning.

@shentao, in your last reply did you mean something similar to my Example 2? Or were you talking about another approach?

Thanks a lot for your time and patience.

**Comment by shentao** [comment]
Hey @eleftrik!

What I meant was something similar to the 2nd example, with the difference that you don’t pass the ajax response up to the container but simply update the options inside the my-list-template.

Here an example of what I mean:
https://jsfiddle.net/2187vzhe/

However, makes it not possible for the parent to know the `cities` list. But if you don’t need them, then I guess that’s fine. As you said – the ajax logic there is always the same anyway.

Happy to see it working for you!

**Comment by eleftrik** [comment]
@shentao Yes, some hour ago I refactored my code, it's pretty similar to your last fiddle.
Actually, when I use Ajax, I don't need to pass the response up to the container.
In my wrapper component I also added some extra params, because I can choose between passing static values (which I receive from the container) or the Ajax way. In this case, options is a simple property updated internally.
Thanks a lot for your time and clarifications!
